example-todomvc/src/main/scala/mhtml/todo/MhtmlTodo.scala

```diff
@@ -0,0 +1,411 @@
+
+case class TodoList(text: String, hash: String, items: Rx[List[Todo]])
+
+sealed trait TodoEvent
+final case class UpdateEvent(oldTodo: Todo, newTodo: Todo) extends TodoEvent
+final case class AddEvent(newTodo: Todo) extends TodoEvent
+final case class RemovalEvent(todo: Todo) extends TodoEvent
+
+
+object MhtmlTodo extends JSApp {
+  type TodoItem = Component[Option[TodoEvent]]
+  type Store = Map[Todo, TodoItem]
+  val allTodosProxy: Var[List[Todo]] = Var(Nil)
+
+  val all = TodoList("All", "#/", allTodosProxy.dropRepeats)
+  val active = TodoList("Active", "#/active", allTodosProxy.map(_.filter(!_.completed)).dropRepeats)
+  val completed = TodoList("Completed", "#/completed", allTodosProxy.map(_.filter(_.completed)).dropRepeats)
+  val todoLists = List(all, active, completed)
+
+  val windowHash: Rx[String] = Rx(dom.window.location.hash).merge {
+    val updatedHash = Var(dom.window.location.hash)
+    dom.window.onhashchange = (ev: Event) => {
+      updatedHash := dom.window.location.hash
+    }
+    updatedHash
+  }.dropRepeats
+
+  val currentTodoList: Rx[TodoList] = windowHash.map { hash =>
+    todoLists.find(_.hash === hash).getOrElse(all)
+  }.dropRepeats
+
+  val todoListEventProxy: Var[Option[TodoEvent]] = Var(None)
+  val todoListComponents: ComponentList[TodoItem, Todo] = {
+
+    val todoListTodos: Rx[List[Todo]] = currentTodoList.map(tl => tl.items).flatten.dropRepeats
+    type TLCMonitors = (TodoList, List[Todo], Option[TodoEvent])
+    val ctlTodosZipped: Rx[TLCMonitors] =
+      (for (ctl <- currentTodoList; tlt <- todoListTodos; tle <- todoListEventProxy)
+        yield {
+          //DEBUG
+          println(s"yield ctl = $ctl}")
+          println(s"yield tlt = $tlt}")
+          println(s"yield tle = $tle}")
+          (ctl, tlt, tle)
+        }).dropRepeats
+
+    val store: Rx[Store] = ctlTodosZipped.foldp[Rx[Store]]( Rx(Map()) ) {
+      (storeNow: Rx[Store], tlcMonitors: TLCMonitors) =>
+        //FIXME: do we really need ev???
+        val (currentTodoList, currentTodos, ev) = tlcMonitors
+        (storeNow |@| currentTodoList.items).map {
+          (currentComps: Store, currentTodos: List[Todo]) =>
+            currentTodos.map { todo =>
+              println(s"currentComps size is ${currentComps.size}")
+              currentComps.get(todo) match {
+                case Some(todoComp) => todo -> todoComp
+                case None => todo -> todoItem(todo)
+              }
+            }
+        }.map[Store]{mapList: List[(Todo, TodoItem)] => mapList.toMap[Todo, TodoItem]}
+    }.flatten.dropRepeats.map{tlc => println(s"got todoListComponents change (predrop)"); tlc}.map{tlc => println(s"got todoListComponents change"); tlc} // DEBUG
+
+    val view: Rx[Node] = store.map{ compMap: Store =>
+      Group(compMap.values.map(item => item.view) (breakOut) )
+    }.dropRepeats
+
+    ComponentList(view, store)
+  }
+
+  //DEBUG
+  todoListComponents.store.impure.foreach(tlc => println(s"tlc size = ${tlc.size}"))
```

Comment:
I basically get that, but agree it should be removed from the code, since aside from usually not showing something useful related to the real Rx, it might have unintended side effects. Still, I'm not quite understanding something at least:

What are the "tons of callbacks"? I thought there is just one - the anonymous `println` call.

Reply:
In general you end up with one callback in [subscriber array](https://github.com/OlivierBlanvillain/monadic-html/blob/master/monadic-rx/shared/src/main/scala/mhtml/rx.scala#L265) of each leaf in your Rx graph. Maybe it does not apply to this example, but given that you have many of those I wouldn't be surprised if the leafs end up with +5 subscribers instead of one.

Reply:
ahh .... I see, and I think I begin to see why `foreach` is worse than I thought, it isn't just memory leaks but execution leaks.